### PR TITLE
Reduce deposit memory

### LINKED ISSUE #4604: Can we reduce memory usage of deposits?
Currently, we keep all of the chain deposits in memory from start up. In my local node, this was around 130Mb. While not a significant memory impact at ~32k deposits, this will become problematic as the network scales.

Screenshot from 2020-01-20 21-07-54

### PR DESCRIPTION
**Pull request by rkapka**
**What type of PR is this?**

Optimization

**What does this PR do? Why is it needed?**

Removes deposit proofs upon finalization. The only place where proofs are needed is verification while performing state transitions. After that they can be safely removed.

At the time of writing this optimization reduced Medalla deposits' memory footprint on my machine from 223.17MB to 19.50MB, which is a 92% improvement.

**Which issues(s) does this PR fix?**

Fixes #4604

**Other notes for review**

Proofs are not removed during initial sync, which means that memory usage reduction will happen after the chain has been synced to the head.

### CODE REVIEW

File: beacon-chain/cache/depositcache/deposits_cache.go [hunk 1]
```diff
@@ -258,3 +258,23 @@ func (dc *DepositCache) NonFinalizedDeposits(ctx context.Context, untilBlk *big.
 
 	return deposits
 }
+
+// PruneProofs removes proofs from all deposits whose index is equal or less than untilDepositIndex.
+func (dc *DepositCache) PruneProofs(ctx context.Context, untilDepositIndex int64) {
+	ctx, span := trace.StartSpan(ctx, "DepositsCache.PruneProofs")
+	defer span.End()
+	dc.depositsLock.RLock()
```
Comment (terencechain) [inline beacon-chain/cache/depositcache/deposits_cache.go:266]: Are you sure read lock is appropriate? I see write action below:

```go
dc.deposits[i].Deposit.Proof = nil
```

File: beacon-chain/cache/depositcache/deposits_cache.go [hunk 2]
```diff
@@ -258,3 +258,23 @@ func (dc *DepositCache) NonFinalizedDeposits(ctx context.Context, untilBlk *big.
...
+	defer dc.depositsLock.RUnlock()
+
+	if untilDepositIndex > int64(len(dc.deposits)) {
+		untilDepositIndex = int64(len(dc.deposits) - 1)
+	}
+
+	for i := untilDepositIndex; i >= 0; i-- {
```
Comment (terencechain) [inline beacon-chain/cache/depositcache/deposits_cache.go:273]: I'm slightly worried this for loop can go on for a while in catastrophic case. Better to add this imo:
```go
if ctx.Err() != nil {
		return nil, ctx.Err()
	}
```

File: beacon-chain/blockchain/process_block.go [hunk 3]
```diff
@@ -125,7 +125,12 @@ func (s *Service) onBlock(ctx context.Context, signed *ethpb.SignedBeaconBlock,
 		if err != nil {
 			return errors.Wrap(err, "could not fetch finalized state")
 		}
-		s.depositCache.InsertFinalizedDeposits(ctx, int64(finalizedState.Eth1Data().DepositCount-1))
+		eth1Index := int64(finalizedState.Eth1Data().DepositCount - 1)
```
Comment (terencechain) [inline beacon-chain/blockchain/process_block.go:128]: Will be nice to include a comment to explain how this eth1 index is derived. 
Reading this is not very clear on how deposit count translates to the index itself.

**Review by terencechain — CHANGES_REQUESTED** [review]
One minor concern on the usage of read lock

**Review by terencechain — APPROVED** [review]

**Comment by rkapka** [comment]
Marking as `OK to merge` since it has been already approved before.

**Review by terencechain — APPROVED** [review]

**Pull request merged**